ports/esp32/main.c

```diff
@@ -214,11 +215,39 @@ void boardctrl_startup(void) {
         nvs_flash_erase();
         nvs_flash_init();
     }
+
+    // Query the physical size of the SPI flash and store it in the size
+    // variable of the global, default SPI flash handle.
+    esp_flash_get_physical_size(NULL, &esp_flash_default_chip->size);
+
+    // If there is no filesystem partition (no "vfs" or "ffat"), add a "vfs" partition
+    // that extends from the end of the application partition up to the end of flash.
+    if (esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_ANY, "vfs") == NULL
+        && esp_partition_find_first(ESP_PARTITION_TYPE_DATA, ESP_PARTITION_SUBTYPE_ANY, "ffat") == NULL) {
+        // No "vfs" or "ffat" partition, so try to create one.
+
+        // Try to find the application partition, to work out where to start the filesystem.
+        const esp_partition_t *part_factory = esp_partition_find_first(ESP_PARTITION_TYPE_APP, ESP_PARTITION_SUBTYPE_ANY, "factory");
+        const esp_partition_t *part_ota_1 = esp_partition_find_first(ESP_PARTITION_TYPE_APP, ESP_PARTITION_SUBTYPE_ANY, "ota_1");
+        size_t offset = 0;
+        if (part_factory != NULL && part_ota_1 == NULL) {
+            offset = part_factory->address + part_factory->size;
+        } else if (part_factory == NULL && part_ota_1 != NULL) {
+            offset = part_ota_1->address + part_ota_1->size;
+        }
```

Comment:
I don't know if it's worth worrying about, but it might be more robust for custom layouts to relax this requirement to "after the last partition in the flash". i.e. (untested):

```c
// Find the end of the last existing partition
size_t offset = 0
esp_partition_iterator_t it = esp_partition_find(ESP_PARTITION_TYPE_ANY, ESP_PARTITION_SUBTYPE_ANY, NULL);
while (it != NULL) {
  const esp_partition_t *part = esp_partition_get(it);
  offset = MAX(offset, part->address + part->size);
  it = esp_partition_next(it);
}
```

Reply:
Yes, that's a good idea, better than what I have here.  It will also be slightly faster because then it only iterates once through the partitions (which requires an alloc in the IDF...).

Your suggestion will also fix a bug in the current code in that it doesn't respect the "romfs" partition and will overwrite it :grimacing:

Reply:
Now updated, and tested with ESP32_GENERIC firmware on a 4MiB board.